### PR TITLE
Refactor!: normalize environment names according to the engine's dialect

### PR DESCRIPTION
**Pull request by georgesittas**
We previously treated environment names as case-insensitive, lowercasing them regardless of the target engine.

This was problematic for engines like Snowflake, because it essentially forced users to use quotes in queries that referenced tables within these environments. So, for example instead of FOO_DEV, we'd create FOO_dev.

This PR aims to address this issue by normalizing the environment names (& the `default` schema) according to the rules of the targeted engine's dialect. So, for the above example users will now be able to query FOO_DEV.table without using quotes for the schema part.

This PR should not affect existing environments, which is achieved by setting their `normalize_name` column to `FALSE`. Hence that column will only be `TRUE` for new environments.

### CODE REVIEW

File: sqlmesh/core/snapshot/definition.py [hunk 1]
```diff
@@ -1291,10 +1314,12 @@ def table_name(
     return exp.table_name(table)
 
 
+# TODO: should we find a way to propagate the engine dialect into `for_environment` here?
```
Comment (georgesittas) [inline sqlmesh/core/snapshot/definition.py:1317]:
```suggestion
```

File: sqlmesh/core/plan/builder.py [hunk 2]
```diff
@@ -137,6 +137,7 @@ def __init__(
             environment_catalog_mapping or {},
             name=self._context_diff.environment,
             suffix_target=environment_suffix_target,
+            normalize_name=self._context_diff.is_new_environment,
```
Comment (tobymao) [inline sqlmesh/core/plan/builder.py:140]: is this for backwards compatibility? how does this work?

File: sqlmesh/migrations/v0052_add_normalize_name_in_environment_naming_info.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+from sqlglot import exp
+
+
+def migrate(state_sync, **kwargs):  # type: ignore
+    engine_adapter = state_sync.engine_adapter
+    environments_table = "_environments"
+    if state_sync.schema:
+        environments_table = f"{state_sync.schema}.{environments_table}"
+
+    alter_table_exp = exp.AlterTable(
```
Comment (tobymao) [inline sqlmesh/migrations/v0052_add_normalize_name_in_environment_naming_info.py:12]: in the future, we should make a builder for alter table, low pri

File: sqlmesh/core/snapshot/definition.py [hunk 4]
```diff
@@ -213,22 +219,36 @@ def catalog_for_environment(
...
         if (
             environment_naming_info.name.lower() != c.PROD
             and environment_naming_info.suffix_target.is_schema
         ):
-            schema = f"{schema}__{environment_naming_info.name}"
+            env_name = environment_naming_info.name
+            if environment_naming_info.normalize_name:
+                env_name = normalize_identifiers(env_name, dialect=dialect).name
+
+            schema = f"{schema}__{env_name}"
```
Comment (tobymao) [inline sqlmesh/core/snapshot/definition.py:234]: how does this work? do we not normalize schema? doesn't that mean it could be

sqlmesh_DEV ?

Comment (tobymao) [reply sqlmesh/core/snapshot/definition.py:234]: or default_DEV

File: sqlmesh/core/plan/builder.py (hunk 2 above)
Comment (georgesittas) [reply sqlmesh/core/plan/builder.py:140]: Correct, new environments should be normalized but old ones shouldn't.

File: sqlmesh/migrations/v0052_add_normalize_name_in_environment_naming_info.py (hunk 3 above)
Comment (georgesittas) [reply sqlmesh/migrations/v0052_add_normalize_name_in_environment_naming_info.py:12]: yep 👍

File: sqlmesh/core/snapshot/definition.py (hunk 4 above)
Comment (georgesittas) [reply sqlmesh/core/snapshot/definition.py:234]: Ah thanks, I think this is an oversight. I'll do another pass at this soon and make sure we normalize everything.

Comment (georgesittas) [reply sqlmesh/core/snapshot/definition.py:234]: I think we should only normalize the `default` schema, right? If a user's set their own schema, then that should have been normalized prior to flowing into this method. I verified this in my tests because `sqlmesh plan foo` ended up creating `FOO__DEV`.

**Comment by georgesittas** [comment]
I plan to wrap this up soon. Remaining items on my end are:

- [ ] Normalize default schema name ([relevant discussion](https://github.com/TobikoData/sqlmesh/pull/2722#discussion_r1626847339))
- [ ] Add tests

File: sqlmesh/core/plan/builder.py (hunk 2 above)
Comment (tobymao) [reply sqlmesh/core/plan/builder.py:140]: why aren't old environments normalized? is this permanent?

Comment (georgesittas) [reply sqlmesh/core/plan/builder.py:140]: I think this may be incorrect after all, thanks for double checking it. I'll revisit.

File: sqlmesh/core/snapshot/definition.py (hunk 4 above)
Comment (georgesittas) [reply sqlmesh/core/snapshot/definition.py:243]: Addressed

File: sqlmesh/core/plan/builder.py (hunk 2 above)
Comment (georgesittas) [reply sqlmesh/core/plan/builder.py]: Addressed a bit differently, see https://github.com/TobikoData/sqlmesh/pull/2722/commits/9ebcc31f5f2b75addda9c12af838a29cea0e2bd1.

TL;DR added new field (`normalize_environment_name: bool`) in `ContextDiff` which is now set as:

```
is_new_environment or bool(env and env.normalize_name)
```

So, if an environment is new, we use the new logic. Otherwise, we check if the fetched environment's `normalize_name` column is `TRUE`.

**Comment by georgesittas** [comment]
This is ready for another review. I added a [simple test](https://github.com/TobikoData/sqlmesh/pull/2722/commits/9ebcc31f5f2b75addda9c12af838a29cea0e2bd1#diff-5af59db99ac14709aa0e02acc49ee767af92605439b2f23c567b4c7d0618b2b1R288-R310) that demonstrates how the default schema & environment name get normalized, but I haven't tested that old environment names won't be normalized (though I did test against snowflake and it seems to work fine). Any suggestions on how we should do this, if needed?

File: sqlmesh/core/console.py [hunk 5]
```diff
@@ -251,6 +251,8 @@ def __init__(
 
         self.verbose = verbose
 
+        self.dialect = kwargs.pop("dialect", None)
```
Comment (izeigerman) [inline sqlmesh/core/console.py:254]: Should we just make this an argument?

File: sqlmesh/core/console.py [hunk 6]
```diff
@@ -1094,6 +1100,8 @@ def __init__(
         self.missing_dates_output = widgets.Output()
         self.dynamic_options_after_categorization_output = widgets.VBox()
 
+        self.dialect = kwargs.pop("dialect", None)
```
Comment (izeigerman) [inline sqlmesh/core/console.py:1103]: ditto

File: sqlmesh/core/context_diff.py [hunk 7]
```diff
@@ -183,10 +186,18 @@ def create_no_diff(cls, environment: str) -> ContextDiff:
         Returns:
             The ContextDiff object.
         """
+        env = state_reader.get_environment(environment.lower())
+
+        if env is None:
+            normalize_environment_name = True
+        else:
+            normalize_environment_name = env.normalize_name
```
Comment (izeigerman) [inline sqlmesh/core/context_diff.py:194]:
```
normalize_environment_name = env is None or not env.normalize_name
```

File: sqlmesh/core/context_diff.py [hunk 8]
```diff
@@ -183,10 +186,18 @@ def create_no_diff(cls, environment: str) -> ContextDiff:
         Returns:
             The ContextDiff object.
         """
+        env = state_reader.get_environment(environment.lower())
```
Comment (izeigerman) [inline sqlmesh/core/context_diff.py:189]: [Nit] I'd probably just pass the environment here instead of the entire state reader.

Comment (izeigerman) [reply sqlmesh/core/context_diff.py:189]: Also makes it easier to test

File: sqlmesh/core/snapshot/definition.py [hunk 9]
```diff
@@ -219,22 +225,39 @@ def catalog_for_environment(
...
         if (
             environment_naming_info.name.lower() != c.PROD
             and environment_naming_info.suffix_target.is_schema
         ):
-            schema = f"{schema}__{environment_naming_info.name}"
+            env_name = environment_naming_info.name
+            if environment_naming_info.normalize_name:
+                env_name = normalize_identifiers(env_name, dialect=dialect).name
+                if not self.schema_name:
+                    # We also normalize the schema if it's the default one
```
Comment (izeigerman) [inline sqlmesh/core/snapshot/definition.py:240]: Why weren't we doing this before? A bug?

Comment (georgesittas) [reply sqlmesh/core/snapshot/definition.py:240]: Yeah, I think so. If we'd set the schema earlier in the control flow, we wouldn't have this issue. But until now, we were basically interpolating a lowercase `default` schema if none was specified, i.e. `"default"."FOO"` would be created in SF.

File: sqlmesh/core/context_diff.py (hunk 8 above)
Comment (georgesittas) [reply sqlmesh/core/context_diff.py:189]: That's a good point, thank you

File: sqlmesh/core/console.py (hunk 5 above)
Comment (georgesittas) [reply sqlmesh/core/console.py:254]: Yep I can do that

**Review by tobymao — APPROVED** [review]

**Review by izeigerman — APPROVED** [review]
Thanks for addressing comments!
